ui/apps/platform/src/Containers/ComplianceEnhanced/ScanConfigs/Wizard/ClusterSelection.tsx

```diff
@@ -1,9 +1,162 @@
+    const { setFieldValue, values: formikValues }: FormikContextType<ScanConfigFormValues> =
+        useFormikContext();
+
+    const clusterIsPreSelected = useCallback(
+        (row) => formikValues.clusters.includes(row.id),
+        [formikValues.clusters]
+    );
+
+    const { allRowsSelected, selected, onSelect, onSelectAll } = useTableSelection(
+        clusters,
+        clusterIsPreSelected
+    );
+
+    useEffect(() => {
+        const clusterIds = clusters.map((cluster) => cluster.id);
+        const selectedClusterIds = clusterIds.filter((_, index) => selected[index]);
+        if (!isEqual(selectedClusterIds, formikValues.clusters)) {
+            setFieldValue('clusters', selectedClusterIds);
+        }
+    }, [selected, formikValues.clusters, setFieldValue, clusters]);
+
+    function renderTableContent() {
+        return clusters?.map(({ id, name }, rowIndex) => (
+            <Tr key={id}>
+                <Td
+                    key={id}
+                    select={{
+                        rowIndex,
+                        onSelect,
+                        isSelected: selected[rowIndex],
+                    }}
+                />
+                <Td>{name}</Td>
+            </Tr>
+        ));
+    }
+
+    function renderLoadingContent() {
+        return (
+            <Tr>
+                <Td colSpan={8}>
+                    <Bullseye>
+                        <Spinner isSVG />
+                    </Bullseye>
+                </Td>
+            </Tr>
+        );
+    }
+
+    function renderEmptyContent() {
+        return (
+            <Tr>
+                <Td colSpan={6}>
+                    <Bullseye>
+                        <EmptyStateTemplate
+                            title="No clusters found"
+                            headingLevel="h2"
+                            icon={SearchIcon}
+                        >
+                            {hasWriteAccessForCluster && (
+                                <Flex direction={{ default: 'column' }}>
+                                    <FlexItem>
+                                        <Text>Install a cluster to get started</Text>
+                                    </FlexItem>
+                                    <FlexItem>
+                                        <InstallClustersButton />
+                                    </FlexItem>
+                                </Flex>
+                            )}
+                        </EmptyStateTemplate>
+                    </Bullseye>
+                </Td>
+            </Tr>
+        );
+    }
+
+    function renderTableBodyContent() {
+        if (isFetchingClusters) {
+            return renderLoadingContent();
+        }
+        if (clusters && clusters.length > 0) {
+            return renderTableContent();
+        }
+        if (clusters && clusters.length === 0) {
+            return renderEmptyContent();
+        }
+        return null;
```

Comment:
What are pro and con of `return renderEmptyContent();` instead of `if` statement and `return null;` here?

Reply:
Here us a pro and con:
* Within `if (clusters && clusters.length > 0) { … }` TypeScript knows that `clusters` is truthy therefore does not need optional chaining `clusters?.map( … )`
* Within `renderTableContent` function, TypeScript does not know when it might be called. Having said that, it looks like the hook guarantees that `clusters` is an array.

Reply:
What I have is more explicit in my opinion, but having it default to renderEmptyContent here might be safer and more concise. I went ahead and updated it in [#8847](https://github.com/stackrox/stackrox/pull/8847)